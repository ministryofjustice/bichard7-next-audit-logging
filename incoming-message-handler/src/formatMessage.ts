import formatMessageXml from "./use-cases/formatMessageXml"
import { clean, hasRootElement } from "./utils/xml"

export default async function formatMessage(event: string): Promise<string> {
  let formattedMessage = event

  const hasDeliveryElement = await hasRootElement(formattedMessage, "DeliverRequest")
  if (!hasDeliveryElement) {
    formattedMessage = formatMessageXml(clean(formattedMessage))
  }

  return formattedMessage
}
