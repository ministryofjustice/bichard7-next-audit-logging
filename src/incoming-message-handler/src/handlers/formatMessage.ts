import { clean, hasRootElement } from "shared"
import type { ReceivedMessage } from "../entities"
import formatMessageXml from "../use-cases/formatMessageXml"

export default async function formatMessage(event: ReceivedMessage): Promise<ReceivedMessage> {
  let formattedMessage = event.messageXml

  formattedMessage = clean(formattedMessage)
  const hasRouteDataElement = await hasRootElement(formattedMessage, "RouteData")

  if (!hasRouteDataElement) {
    formattedMessage = formatMessageXml(formattedMessage)
  }

  return {
    s3Path: event.s3Path,
    externalId: event.externalId,
    stepExecutionId: event.stepExecutionId,
    receivedDate: event.receivedDate,
    messageXml: formattedMessage
  }
}
