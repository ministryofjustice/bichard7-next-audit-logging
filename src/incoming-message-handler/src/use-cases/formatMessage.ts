import { clean, hasRootElement } from "shared"
import type { PromiseResult } from "shared-types"
import type { ReceivedMessage } from "../entities"
import ApplicationError from "src/errors/ApplicationError"
import formatMessageXml from "./formatMessageXml"
import crypto from "crypto"

const generateHash = (text: string) => crypto.createHash("sha256").update(text, "utf-8").digest("hex")

const getInnerMessage = (xml: string) =>
  xml.match(/<(?:[\S]*:)?DataStreamContent>(?<innerMessage>[\s\S]*)<\/(?:[\S]*:)?DataStreamContent>/)?.groups
    ?.innerMessage

export default async (receivedMessage: ReceivedMessage): PromiseResult<ReceivedMessage> => {
  try {
    let { messageXml } = receivedMessage
    messageXml = clean(messageXml)
    const hasRouteDataElement = await hasRootElement(messageXml, "RouteData")

    let hash: string
    if (!hasRouteDataElement) {
      hash = generateHash(messageXml)
      messageXml = formatMessageXml(messageXml)
    } else {
      const innerXml = getInnerMessage(messageXml)
      hash = generateHash(innerXml ?? messageXml)
    }

    return {
      ...receivedMessage,
      messageXml,
      hash
    }
  } catch (error) {
    return new ApplicationError("Error while formatting the message", error as Error)
  }
}
