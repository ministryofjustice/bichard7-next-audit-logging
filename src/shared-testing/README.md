# Audit Log Shared Testing package

A library with shared toools used for testing.
