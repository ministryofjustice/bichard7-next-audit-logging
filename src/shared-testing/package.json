{
  "name": "shared-testing",
  "description": "Provides libraries used for testing",
  "version": "1.0.0",
  "author": "",
  "license": "ISC",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "build/**/*"
  ],
  "scripts": {
    "clean": "rimraf build",
    "build:dev": "npm run clean && tsc -p tsconfig.json",
    "build": "npm run clean && tsc -p tsconfig.prod.json",
    "test": "jest --runInBand",
    "test:unit": "jest",
    "test:unit:watch": "jest --watch"
  },
  "dependencies": {
    "aws-sdk": "^2.1238.0",
    "axios": "^1.1.3",
    "shared-types": "file:../shared-types",
    "uuid": "^9.0.0"
  },
  "devDependencies": {
    "@types/jest": "^29.2.0",
    "@types/uuid": "^8.3.4",
    "jest": "^29.2.1",
    "rimraf": "^3.0.2",
    "ts-jest": "^29.0.3",
    "typescript": "^4.8.4"
  }
}
