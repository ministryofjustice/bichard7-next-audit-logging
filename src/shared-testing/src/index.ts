// eslint-disable-next-line import/prefer-default-export
export { default as FakeApiClient } from "./FakeApiClient"
// eslint-disable-next-line import/prefer-default-export
export { default as setEnvironmentVariables } from "./setEnvironmentVariables"
// eslint-disable-next-line import/prefer-default-export
export { default as FakeAuditLogDynamoGateway } from "./FakeAuditLogDynamoGateway"
// eslint-disable-next-line import/prefer-default-export
export { default as invokeFunction } from "./invokeFunction"
// eslint-disable-next-line import/prefer-default-export
export { default as FakeMqGateway } from "./FakeMqGateway"
// eslint-disable-next-line import/prefer-default-export
export { default as FakeS3Gateway } from "./FakeS3Gateway"
// eslint-disable-next-line import/prefer-default-export
export { default as StepFunctionSimulator } from "./StepFunctionSimulator"
export * from "./createMockAuditLogs"
export * from "./mockAuditLogs"

import "./jest"
