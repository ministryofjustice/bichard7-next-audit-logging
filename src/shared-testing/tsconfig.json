{
  "extends": "../../.config/tsconfig.json",
  "compilerOptions": {
    "outDir": "build",
    "types": [
      "node",
      "jest"
    ],
    "sourceMap": false,
    "baseUrl": ".",
    "preserveSymlinks": false,
    "declaration": true,
    "paths": {
      "src/*": [
        "src/*"
      ]
    }
  },
  "compileOnSave": true,
  "include": [
    "src/**/*.ts"
  ]
}
