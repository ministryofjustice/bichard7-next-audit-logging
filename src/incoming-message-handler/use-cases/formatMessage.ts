import crypto from "crypto"
import { clean, hasRootElement } from "src/shared"
import type { PromiseResult } from "src/shared/types"
import { ApplicationError } from "src/shared/types"
import type { ReceivedMessage } from "../entities"
import formatMessageXml from "./formatMessageXml"
import getDataStreamContent from "./getDataStreamContent"

const generateHash = (text: string) => crypto.createHash("sha256").update(text, "utf-8").digest("hex")

export default async (receivedMessage: ReceivedMessage): PromiseResult<ReceivedMessage> => {
  try {
    let { messageXml } = receivedMessage
    messageXml = clean(messageXml)
    const hasRouteDataElement = await hasRootElement(messageXml, "RouteData")

    let hash: string
    if (!hasRouteDataElement) {
      hash = generateHash(messageXml.trim())
      messageXml = formatMessageXml(messageXml)
    } else {
      const innerXml = getDataStreamContent(messageXml)
      hash = generateHash(innerXml ?? messageXml.trim())
    }

    return {
      ...receivedMessage,
      messageXml,
      hash
    }
  } catch (error) {
    return new ApplicationError("Error while formatting the message", error as Error)
  }
}
