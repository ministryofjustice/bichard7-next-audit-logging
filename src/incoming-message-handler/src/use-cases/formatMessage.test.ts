import "shared-testing"
import type { ReceivedMessage } from "../entities"
import formatMessage from "./formatMessage"

const message = `
<DC:ResultedCaseMessage xmlns:DC="http://www.dca.gov.uk/xmlschemas/libra" Flow='ResultedCasesForThePolice' Interface='LibraStandardProsecutorPolice' SchemaVersion='0.6g'>
	&lt;DC:Session&gt;
		<DC:Case>
			<DC:PTIURN>
        123456789
      </DC:PTIURN>
		</DC:Case>
	&lt;/DC:Session&gt;
</DC:ResultedCaseMessage>
`

it("should format the message XML when the message XML is correct", async () => {
  const receivedMessage = {
    s3Path: "DUMMY_KEY",
    receivedDate: "DUMMY_DATE",
    externalId: "DUMMY_EXTERNALID",
    stepExecutionId: "DUMMY_EXECUTION_ID",
    messageXml: message
  } as ReceivedMessage
  const result = await formatMessage(receivedMessage)

  expect(result).toNotBeError()

  const { messageXml, s3Path, stepExecutionId, externalId, receivedDate } = result as ReceivedMessage
  expect(s3Path).toBe(receivedMessage.s3Path)
  expect(stepExecutionId).toBe(receivedMessage.stepExecutionId)
  expect(externalId).toBe(receivedMessage.externalId)
  expect(receivedDate).toBe(receivedMessage.receivedDate)
  expect(messageXml).toMatchSnapshot()
})

it("should return error when message XML is incorrect", async () => {
  const receivedMessage = {
    s3Path: "DUMMY_KEY",
    receivedDate: "DUMMY_DATE",
    externalId: "DUMMY_EXTERNALID",
    stepExecutionId: "DUMMY_EXECUTION_ID",
    messageXml: "INCORRECT_XML"
  } as ReceivedMessage
  const result = await formatMessage(receivedMessage)

  expect(result).toBeError("Error while formatting the message")
})
