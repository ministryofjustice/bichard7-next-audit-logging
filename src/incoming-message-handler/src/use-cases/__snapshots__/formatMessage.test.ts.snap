// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should format the message XML 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
  <RouteData xmlns=\\"http://schemas.cjse.gov.uk/common/operations\\" xmlns:cjseEntity=\\"http://schemas.cjse.gov.uk/common/businessentities\\" xmlns:cjseType=\\"http://schemas.cjse.gov.uk/common/businesstypes\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" VersionNumber=\\"1.0\\" RequestResponse=\\"Request\\">
    <RequestFromSystem VersionNumber=\\"1.0\\">
      <CorrelationID>2a0d2aed-c26d-41c5-8bef-1e60c657c182</CorrelationID>
    </RequestFromSystem>
    <DataStream VersionNumber=\\"1.0\\">
      <DataStreamContent>
<DC:ResultedCaseMessage xmlns:DC=\\"http://www.dca.gov.uk/xmlschemas/libra\\" Flow='ResultedCasesForThePolice' Interface='LibraStandardProsecutorPolice' SchemaVersion='0.6g'>
	<DC:Session>
		<DC:Case>
			<DC:PTIURN>
        123456789
      </DC:PTIURN>
		</DC:Case>
	</DC:Session>
</DC:ResultedCaseMessage>
</DataStreamContent>
    </DataStream>
  </RouteData>"
`;

exports[`should format the message XML when the message XML is correct 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
  <RouteData xmlns=\\"http://schemas.cjse.gov.uk/common/operations\\" xmlns:cjseEntity=\\"http://schemas.cjse.gov.uk/common/businessentities\\" xmlns:cjseType=\\"http://schemas.cjse.gov.uk/common/businesstypes\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" VersionNumber=\\"1.0\\" RequestResponse=\\"Request\\">
    <RequestFromSystem VersionNumber=\\"1.0\\">
      <CorrelationID>d97733bb-0acc-4f98-a07e-6f413e5b39c2</CorrelationID>
    </RequestFromSystem>
    <DataStream VersionNumber=\\"1.0\\">
      <DataStreamContent>
<DC:ResultedCaseMessage xmlns:DC=\\"http://www.dca.gov.uk/xmlschemas/libra\\" Flow='ResultedCasesForThePolice' Interface='LibraStandardProsecutorPolice' SchemaVersion='0.6g'>
	<DC:Session>
		<DC:Case>
			<DC:PTIURN>
        123456789
      </DC:PTIURN>
		</DC:Case>
	</DC:Session>
</DC:ResultedCaseMessage>
</DataStreamContent>
    </DataStream>
  </RouteData>"
`;
